kythe/extractors/openjdk11/java_wrapper/java_wrapper.go

```diff
@@ -0,0 +1,215 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"bitbucket.org/creachadair/shell"
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	javaCommandVar   = "KYTHE_JAVA_COMMAND"     // Path to the real java command, required.
+	kytheRootVar     = "KYTHE_ROOT_DIRECTORY"   // Extraction root directory, set to "." if unspecified.
+	kytheOutputVar   = "KYTHE_OUTPUT_DIRECTORY" // Extraction output directory, defaults to ~/kythe-openjdk11-output
+	kytheCorpusVar   = "KYTHE_CORPUS"           // Extraction corpus to use, if KYTHE_VNAMES is unset. Defaults to openjdk11.
+	kytheTargetVar   = "KYTHE_ANALYSIS_TARGET"  // Extraction analysis target to set, defaults to the java module name.
+	extractorJarVar  = "JAVAC_EXTRACTOR_JAR"    // Path to the javac_extractor jar, default to Bazel runfiles.
+	extractorJarPath = "io_kythe/kythe/java/com/google/devtools/kythe/extractors/java/standalone/javac9_extractor_deploy.jar"
+)
+
+var (
+	modulePattern = regexp.MustCompile("^@/.*/_the.(.*)_batch.tmp$")
+)
+
+func moduleName() string {
+	// Hackish way to determine the likely module being compiled.
+	return modulePattern.ReplaceAllString(os.Args[len(os.Args)-1], "$1")
+}
+
+func outputDir() string {
+	val := os.Getenv(kytheOutputVar)
+	if len(val) == 0 {
+		usr, err := user.Current()
+		if err != nil {
+			log.Fatalf("ERROR: unable to determine current user: %v", err)
+		}
+		val = filepath.Join(usr.HomeDir, "kythe-openjdk11-output")
+	}
+	return val
+}
+
+func checkPath(parts ...string) error {
+	_, err := os.Stat(filepath.Join(parts...))
+	return err
+}
+
+func setupRunfiles() error {
+	if os.Getenv("RUNFILES_DIR") != "" || os.Getenv("RUNFILES_MANIFEST_FILE") != "" {
+		return nil
+	}
+	for _, base := range []string{os.Args[0] + ".runfiles", "."} {
+		root, err := filepath.Abs(base)
+		if err != nil {
+			continue
+		} else if _, err := os.Stat(root); err != nil {
+			continue
+		}
+		os.Setenv("RUNFILES_DIR", root)
+		manifest := filepath.Join(root, "MANIFEST")
+		if _, err := os.Stat(manifest); err == nil {
+			os.Setenv("RUNFILES_MANIFEST_FILE", manifest)
+		}
+		return nil
+	}
+	return errors.New("unable to setup runfiles")
+}
+
+func findJavaCommand() (string, error) {
+	if val := os.Getenv(javaCommandVar); len(val) > 0 {
+		return val, checkPath(val)
+	}
+	return "", os.ErrNotExist
+}
+
+func findExtractorJar() (string, error) {
+	if val := os.Getenv(extractorJarVar); len(val) > 0 {
+		return val, checkPath(val)
+	}
+	return bazel.Runfile(extractorJarPath)
+}
+
+func extractorArgs(args []string, jar string) []string {
+	isJavac := false
+	var result []string
+	for len(args) > 0 {
+		var a string
+		var v string
+		switch a, args = shift(args); a {
+		case "-m", "--module":
+			v, args = shift(args)
+			if !strings.HasSuffix(v, ".javac.Main") {
+				isJavac = false
```

Comment:
Could we just `return nil` at this point and avoid the need for a flag var? It looks like the only place we check this is at the end anyway.

Reply:
We cannot, sadly as not all of the non-javac invocations have a "-m" flag at all.

Reply:
Ah. So we could return here, but we would still need the flag. Boo. 😢